ares_init.c

```diff
@@ -1845,6 +1860,87 @@ static int init_by_resolv_conf(ares_channel channel)
   return ARES_SUCCESS;
 }
 
+int ares_reinit_by_resolv_conf_file(ares_channel channel)
+{
+#if defined(ANDROID) || defined(__ANDROID__) || defined(WIN32) || defined(WATT32) || \
+    defined(CARES_USE_LIBRESOLV) || defined(__riscos__)
+
+    return -1;
+
+#else
+
+  struct stat st;
+  const char *resolvconf_path;
+  if (channel->resolvconf_path) {
+    resolvconf_path = channel->resolvconf_path;
+  }
+  else {
+    resolvconf_path = PATH_RESOLV_CONF;
+  }
+  if (stat(resolvconf_path, &st) != 0) {
+      return ARES_ENODATA;
+  }
+
+  /* If the stat values are the same, we do not want to reinit */
+#ifdef __APPLE__
+  if (channel->resolvconf_mtime.tv_sec == st.st_mtimespec.tv_sec
+      && channel->resolvconf_mtime.tv_nsec == st.st_mtimespec.tv_nsec
+      && channel->resolvconf_ctime.tv_sec == st.st_ctimespec.tv_sec
+      && channel->resolvconf_ctime.tv_nsec == st.st_ctimespec.tv_nsec
+#else
+  if (channel->resolvconf_mtime.tv_sec == st.st_mtim.tv_sec
+      && channel->resolvconf_mtime.tv_nsec == st.st_mtim.tv_nsec
+      && channel->resolvconf_ctime.tv_sec == st.st_ctim.tv_sec
+      && channel->resolvconf_ctime.tv_nsec == st.st_ctim.tv_nsec
+#endif
+      && channel->resolvconf_ino == st.st_ino
+      && channel->resolvconf_size == st.st_size) {
+    return ARES_SUCCESS;
+  }
+
+  if (channel->ndomains != -1) {
+    ares_strsplit_free(channel->domains, channel->ndomains);
+    channel->domains = NULL;
+    channel->ndomains = -1;
+  }
+
+  if (channel->lookups) {
+    ares_free(channel->lookups);
+    channel->lookups = NULL;
+  }
+
+  /* if we have channel server created,
+     we need to also disconnect and close sockets */
+  if (channel->nservers != -1) {
+    ares__destroy_servers_state(channel);
+  }
+
+  if (channel->sortlist) {
+    ares_free(channel->sortlist);
+  }
+  channel->nsort = -1;
+
+  /* clear options */
+  channel->ndots = -1;
+  channel->timeout = -1;
+  channel->tries = -1;
+  channel->rotate = -1;
+
+  int rc = init_by_resolv_conf(channel);
+  if (init_by_resolv_conf(channel) != ARES_SUCCESS) {
```

Comment:
Why `init_by_resolv_conf` is called twice here?

Reply:
Sorry, that is is a typo.

Reply:
I drafted a patch and attempt to make this part more transactional, i.e., don't update timestamp unless `init_by_resolv_conf` succeeds.
I'm not sure whether that can resolve the crash. In theory, if the first call in the typo succeeds and the second call fails, `servers[0]->qhead` could be uninitialized while `nservers > 0`, but this is only my guess.